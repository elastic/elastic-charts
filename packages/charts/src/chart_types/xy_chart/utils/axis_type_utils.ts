/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { Position } from '../../../utils/common';

/** @internal */
export function isVerticalAxis(axisPosition: Position): axisPosition is Extract<Position, 'left' | 'right'> {
  return axisPosition === Position.Left || axisPosition === Position.Right;
}

/** @internal */
export function isHorizontalAxis(axisPosition: Position): axisPosition is Extract<Position, 'top' | 'bottom'> {
  return axisPosition === Position.Top || axisPosition === Position.Bottom;
}

/** @internal */
export function isVerticalGrid(axisPosition: Position) {
  return isHorizontalAxis(axisPosition);
}

/** @internal */
export function isHorizontalGrid(axisPosition: Position) {
  return isVerticalAxis(axisPosition);
}
