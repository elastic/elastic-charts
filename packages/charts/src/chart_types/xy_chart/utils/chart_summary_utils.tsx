/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import type { BasicSeriesSpec } from './specs';
import { capitalizeFirst } from '../../../utils/text/text_utils';
import type { SeriesDomainsAndData } from '../state/utils/types';

function getDataSummary(seriesDomains: SeriesDomainsAndData): string {
  const dataCount = seriesDomains.formattedDataSeries[0]?.data?.length || 0;
  const xDomain = seriesDomains.xDomain;

  if (dataCount === 0) return '';

  if (xDomain?.type === 'ordinal') {
    return `with ${dataCount} ${dataCount === 1 ? 'category' : 'categories'}`;
  } else if (xDomain?.type === 'time') {
    return `with ${dataCount} time ${dataCount === 1 ? 'period' : 'periods'}`;
  }
  return `with ${dataCount} data ${dataCount === 1 ? 'point' : 'points'}`;
}

function getValueRangeContext(seriesDomains: SeriesDomainsAndData): string {
  const firstSeries = seriesDomains.formattedDataSeries[0];
  if (!firstSeries?.data || firstSeries.data.length === 0) {
    return '';
  }

  // Get actual data values, not the domain (which may include 0 for bar charts)
  const yValues = firstSeries.data.map((d) => d.y1 ?? d.y0 ?? 0).filter((val) => val !== null);
  if (yValues.length === 0) {
    return '';
  }

  const min = Math.min(...yValues);
  const max = Math.max(...yValues);

  if (min === max) {
    if (yValues.length === 1) {
      return `, value is ${min}`;
    }
    return `, all values are ${min}`;
  }
  return `, values ranging from ${min} to ${max}`;
}

/** @internal */
export function createChartTypeDescription(
  chartTypeDescription: string,
  seriesSpecs?: BasicSeriesSpec[],
  seriesDomains?: SeriesDomainsAndData,
): string {
  if (!chartTypeDescription || !seriesSpecs || !seriesDomains?.formattedDataSeries) {
    return chartTypeDescription;
  }

  const seriesTypes = new Set<string>();
  seriesSpecs.forEach((spec) => seriesTypes.add(spec.seriesType));

  const actualSeriesCount = seriesDomains.formattedDataSeries.length;
  const renderedSeries = seriesDomains.formattedDataSeries;

  const seriesNames = renderedSeries
    .map((series) => {
      const spec = seriesSpecs.find((s) => s.id === series.specId);
      if (spec && typeof spec.name === 'string') return spec.name;
      if (series.splitAccessors && series.splitAccessors.size > 0) {
        const splitValues = Array.from(series.splitAccessors.values());
        return splitValues.join(' ');
      }
      return `Series ${series.specId}`;
    })
    .filter((name) => name);

  const hasStackedSeries = seriesSpecs.some(
    (spec) => 'stackAccessors' in spec && spec.stackAccessors && spec.stackAccessors.length > 0,
  );

  if (seriesTypes.size === 1) {
    const seriesType = Array.from(seriesTypes)[0];

    const hasPercentageStacking = seriesSpecs.some((spec) => 'stackMode' in spec && spec.stackMode === 'percentage');

    const stackPrefix = hasStackedSeries ? (hasPercentageStacking ? 'percentage stacked' : 'stacked') : '';

    if (actualSeriesCount === 1) {
      const dataSummary = seriesDomains ? getDataSummary(seriesDomains) : '';
      const valueRange = seriesDomains ? getValueRangeContext(seriesDomains) : '';
      const contextInfo = dataSummary + valueRange;
      return capitalizeFirst(
        `${stackPrefix ? `${stackPrefix} ` : ''}${seriesType} chart${contextInfo ? ` ${contextInfo}` : ''}`,
      );
    } else {
      const chartTypeDescriptionStackChecked = `${stackPrefix ? `${stackPrefix} ` : ''}${seriesType} chart`;
      const countDescription = `with ${actualSeriesCount} ${seriesType}s`;

      const description = `${chartTypeDescriptionStackChecked} ${countDescription}`;
      const finalDescription =
        seriesNames.length > 0 && seriesNames.length <= 5 ? `${description}: ${seriesNames.join(', ')}` : description;
      return capitalizeFirst(finalDescription);
    }
  } else {
    const stackPrefix = hasStackedSeries ? 'stacked mixed' : 'mixed';
    return capitalizeFirst(`${stackPrefix} chart: ${Array.from(seriesTypes).join(' and ')} chart`);
  }
}
