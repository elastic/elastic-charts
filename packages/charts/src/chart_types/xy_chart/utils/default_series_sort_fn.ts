/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { DataSeries } from './series';

/**
 * Return the default sorting used for XY series.
 * Ordered by group insert order, then first stacked, after non stacked.
 * @internal
 */
export function defaultXYSeriesSort(a: DataSeries, b: DataSeries) {
  if (a.groupId !== b.groupId) {
    return a.insertIndex - b.insertIndex;
  }

  if (a.isStacked && !b.isStacked) {
    return -1; // a first then b
  }
  if (!a.isStacked && b.isStacked) {
    return 1; // b first then a
  }
  return a.insertIndex - b.insertIndex;
}

/**
 * Return the default sorting used for XY series.
 * Ordered by group insert order, then first stacked, after non stacked.
 * Stacked are sorted from from top to bottom to respect the rendering order
 * @internal
 */
export function defaultXYLegendSeriesSort(a: DataSeries, b: DataSeries) {
  if (a.groupId !== b.groupId) {
    return a.insertIndex - b.insertIndex;
  }

  if (a.isStacked && !b.isStacked) {
    return -1; // a first then b
  }
  if (!a.isStacked && b.isStacked) {
    return 1; // b first then a
  }
  if (a.isStacked && b.isStacked) {
    return b.insertIndex - a.insertIndex;
  }
  return a.insertIndex - b.insertIndex;
}
