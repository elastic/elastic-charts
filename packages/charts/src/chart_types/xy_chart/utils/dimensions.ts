/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { AxisSpec } from './specs';
import { SmallMultiplesSpec } from '../../../specs';
import { Dimensions } from '../../../utils/dimensions';
import { AxisId } from '../../../utils/ids';
import { Theme, AxisStyle } from '../../../utils/themes/theme';
import { getAxesDimensions } from '../axes/axes_sizes';
import { AxesTicksDimensions } from '../state/selectors/compute_axis_ticks_dimensions';

/**
 * @internal
 */
export interface ChartDimensions {
  /**
   * Dimensions relative to canvas element
   */
  chartDimensions: Dimensions;
  /**
   * Margin to account for ending text overflow
   */
  leftMargin: number;
} /**/

/**
 * Compute the chart dimensions. It's computed removing from the parent dimensions
 * the axis spaces, the legend and any other specified style margin and padding.
 * @internal
 */ export function computeChartDimensions(
  parentDimensions: Dimensions,
  theme: Theme,
  axisTickDimensions: AxesTicksDimensions,
  axesStyles: Map<AxisId, AxisStyle | null>,
  axisSpecs: AxisSpec[],
  smSpec: SmallMultiplesSpec | null,
): ChartDimensions {
  const axesDimensions = getAxesDimensions(parentDimensions, theme, axisTickDimensions, axesStyles, axisSpecs, smSpec);
  const chartWidth = parentDimensions.width - axesDimensions.left - axesDimensions.right;
  const chartHeight = parentDimensions.height - axesDimensions.top - axesDimensions.bottom;
  const pad = theme.chartPaddings;
  return {
    leftMargin: axesDimensions.margin.left,
    chartDimensions: {
      top: axesDimensions.top + pad.top,
      left: axesDimensions.left + pad.left,
      width: Math.max(0, chartWidth - pad.left - pad.right),
      height: Math.max(0, chartHeight - pad.top - pad.bottom),
    },
  };
}
