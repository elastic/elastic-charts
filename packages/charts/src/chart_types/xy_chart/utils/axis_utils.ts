/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { isHorizontalAxis, isVerticalAxis } from './axis_type_utils';
import { computeXScale, computeYScales } from './scales';
import { SmallMultipleScales, hasSMDomain, getPanelSize } from '../../../common/panel_utils';
import { ScaleBand, ScaleContinuous } from '../../../scales';
import { isContinuousScale } from '../../../scales/types';
import { AxisSpec, SettingsSpec } from '../../../specs';
import {
  degToRad,
  getPercentageValue,
  HorizontalAlignment,
  Position,
  Rotation,
  VerticalAlignment,
} from '../../../utils/common';
import { Dimensions, innerPad, outerPad, PerSideDistance, Size } from '../../../utils/dimensions';
import { Range } from '../../../utils/domain';
import { AxisId } from '../../../utils/ids';
import { Point } from '../../../utils/point';
import { AxisStyle, TextAlignment, TextOffset, Theme } from '../../../utils/themes/theme';
import { Projection } from '../state/selectors/visible_ticks';
import { SeriesDomainsAndData } from '../state/utils/types';

type TickValue = number | string;

/** @internal */
export type TextDirection = 'rtl' | 'ltr';

/** @internal */
export interface AxisTick {
  value: TickValue;
  domainClampedValue: TickValue;
  label: string;
  position: number;
  domainClampedPosition: number;
  layer?: number;
  detailedLayer: number;
  showGrid: boolean;
  direction: TextDirection;
}

/** @internal */
export interface TickLabelBounds {
  maxLabelBboxWidth: number;
  maxLabelBboxHeight: number;
  maxLabelTextWidth: number;
  maxLabelTextHeight: number;
  isHidden: boolean;
}

interface TickLabelProps {
  x: number;
  y: number;
  offsetX: number;
  offsetY: number;
  textOffsetX: number;
  textOffsetY: number;
  horizontalAlign: Extract<
    HorizontalAlignment,
    typeof HorizontalAlignment.Left | typeof HorizontalAlignment.Center | typeof HorizontalAlignment.Right
  >;
  verticalAlign: Extract<
    VerticalAlignment,
    typeof VerticalAlignment.Top | typeof VerticalAlignment.Middle | typeof VerticalAlignment.Bottom
  >;
}

/** @internal */
export const defaultTickFormatter = (tick: unknown) => `${tick}`;

/** @internal */
export function isXDomain(position: Position, chartRotation: Rotation): boolean {
  return isHorizontalAxis(position) === (chartRotation % 180 === 0);
}

/** @internal */
export function getScaleForAxisSpec(
  { xDomain, yDomains }: Pick<SeriesDomainsAndData, 'xDomain' | 'yDomains'>,
  { rotation: chartRotation }: Pick<SettingsSpec, 'rotation'>,
  totalBarsInCluster: number,
  barsPadding?: number,
  enableHistogramMode?: boolean,
) {
  return (
    {
      groupId,
      integersOnly,
      maximumFractionDigits: mfd,
      position,
    }: Pick<AxisSpec, 'groupId' | 'integersOnly' | 'maximumFractionDigits' | 'position'>,
    range: Range,
  ): ScaleContinuous | ScaleBand | null => {
    // TODO: remove this fallback when integersOnly is removed
    const maximumFractionDigits = mfd ?? (integersOnly ? 0 : undefined);

    return isXDomain(position, chartRotation)
      ? computeXScale({ xDomain, totalBarsInCluster, range, barsPadding, enableHistogramMode, maximumFractionDigits })
      : computeYScales({ yDomains, range, maximumFractionDigits }).get(groupId) ?? null;
  };
}

/** @internal */
export function computeRotatedLabelDimensions(unrotatedDims: Size, degreesRotation: number): Size {
  const { width, height } = unrotatedDims;
  const radians = degToRad(degreesRotation);
  const rotatedHeight = Math.abs(width * Math.sin(radians)) + Math.abs(height * Math.cos(radians));
  const rotatedWidth = Math.abs(width * Math.cos(radians)) + Math.abs(height * Math.sin(radians));
  return { width: rotatedWidth, height: rotatedHeight };
}

function getUserTextOffsets(dimensions: TickLabelBounds, { x, y, reference }: TextOffset) {
  return reference === 'global'
    ? {
        local: { x: 0, y: 0 },
        global: {
          x: getPercentageValue(x, dimensions.maxLabelBboxWidth, 0),
          y: getPercentageValue(y, dimensions.maxLabelBboxHeight, 0),
        },
      }
    : {
        local: {
          x: getPercentageValue(x, dimensions.maxLabelTextWidth, 0),
          y: getPercentageValue(y, dimensions.maxLabelTextHeight, 0),
        },
        global: { x: 0, y: 0 },
      };
}

const horizontalOffsetMultiplier = {
  [HorizontalAlignment.Left]: -1,
  [HorizontalAlignment.Right]: 1,
  [HorizontalAlignment.Center]: 0,
};

const verticalOffsetMultiplier = {
  [VerticalAlignment.Top]: -1,
  [VerticalAlignment.Bottom]: 1,
  [VerticalAlignment.Middle]: 0,
};

function getHorizontalAlign(
  position: Position,
  rotation: number,
  alignment: HorizontalAlignment,
): Exclude<HorizontalAlignment, typeof HorizontalAlignment.Far | typeof HorizontalAlignment.Near> {
  if (
    alignment === HorizontalAlignment.Center ||
    alignment === HorizontalAlignment.Right ||
    alignment === HorizontalAlignment.Left
  ) {
    return alignment;
  }

  if (Math.abs(rotation) === 90) {
    if (position === Position.Top) {
      return rotation === 90 ? HorizontalAlignment.Right : HorizontalAlignment.Left;
    } else if (position === Position.Bottom) {
      return rotation === -90 ? HorizontalAlignment.Right : HorizontalAlignment.Left;
    }
  } else {
    if (position === Position.Left) {
      return alignment === HorizontalAlignment.Near ? HorizontalAlignment.Right : HorizontalAlignment.Left;
    } else if (position === Position.Right) {
      return alignment === HorizontalAlignment.Near ? HorizontalAlignment.Left : HorizontalAlignment.Right;
    }
  }

  return HorizontalAlignment.Center; // fallback for near/far on top/bottom axis
}

function getVerticalAlign(
  position: Position,
  rotation: number,
  alignment: VerticalAlignment,
): Exclude<VerticalAlignment, typeof VerticalAlignment.Far | typeof VerticalAlignment.Near> {
  if (
    alignment === VerticalAlignment.Middle ||
    alignment === VerticalAlignment.Top ||
    alignment === VerticalAlignment.Bottom
  ) {
    return alignment;
  }

  if (rotation % 180 === 0) {
    if (position === Position.Left) {
      return rotation === 0 ? VerticalAlignment.Bottom : VerticalAlignment.Top;
    } else if (position === Position.Right) {
      return rotation === 180 ? VerticalAlignment.Bottom : VerticalAlignment.Top;
    }
  } else {
    if (position === Position.Top) {
      return alignment === VerticalAlignment.Near ? VerticalAlignment.Bottom : VerticalAlignment.Top;
    } else if (position === Position.Bottom) {
      return alignment === VerticalAlignment.Near ? VerticalAlignment.Top : VerticalAlignment.Bottom;
    }
  }

  return VerticalAlignment.Middle; // fallback for near/far on left/right axis
}

/** @internal */
export function getTickLabelPosition(
  { tickLine, tickLabel }: AxisStyle,
  tickPosition: number,
  pos: Position,
  rotation: number,
  axisSize: Size,
  tickDimensions: TickLabelBounds,
  showTicks: boolean,
  textOffset: TextOffset,
  textAlignment: TextAlignment,
): TickLabelProps {
  const { maxLabelBboxWidth, maxLabelTextWidth, maxLabelBboxHeight, maxLabelTextHeight } = tickDimensions;
  const tickDimension = showTicks ? tickLine.size + tickLine.padding : 0;
  const labelInnerPadding = innerPad(tickLabel.padding);
  const horizontalAlign = getHorizontalAlign(pos, rotation, textAlignment.horizontal);
  const verticalAlign = getVerticalAlign(pos, rotation, textAlignment.vertical);
  const userOffsets = getUserTextOffsets(tickDimensions, textOffset);
  const paddedTickDimension = tickDimension + labelInnerPadding;
  const axisNetSize = (isVerticalAxis(pos) ? axisSize.width : axisSize.height) - paddedTickDimension;
  const labelBoxHalfGirth = isHorizontalAxis(pos) ? maxLabelBboxHeight / 2 : maxLabelBboxWidth / 2;
  const labelHalfWidth = maxLabelTextWidth / 2;
  return {
    horizontalAlign,
    verticalAlign,
    x: pos === Position.Left ? axisNetSize : pos === Position.Right ? paddedTickDimension : tickPosition,
    y: pos === Position.Top ? axisNetSize : pos === Position.Bottom ? paddedTickDimension : tickPosition,
    offsetX: userOffsets.global.x + (isHorizontalAxis(pos) ? 0 : horizontalOffsetMultiplier[pos] * labelBoxHalfGirth),
    offsetY: userOffsets.global.y + (isVerticalAxis(pos) ? 0 : verticalOffsetMultiplier[pos] * labelBoxHalfGirth),
    textOffsetX:
      userOffsets.local.x +
      (isHorizontalAxis(pos) && rotation === 0 ? 0 : labelHalfWidth * horizontalOffsetMultiplier[horizontalAlign]),
    textOffsetY: userOffsets.local.y + (maxLabelTextHeight / 2) * verticalOffsetMultiplier[verticalAlign],
  };
}

/** @internal */
export function getTitleDimension({
  visible,
  fontSize,
  padding,
}: AxisStyle['axisTitle'] | AxisStyle['axisPanelTitle']): number {
  return visible && fontSize > 0 ? innerPad(padding) + fontSize + outerPad(padding) : 0;
}

/** @internal */
export const getAllAxisLayersGirth = (
  timeAxisLayerCount: number,
  maxLabelBoxGirth: number,
  axisHorizontal: boolean,
) => {
  const axisLayerCount = timeAxisLayerCount > 0 && axisHorizontal ? timeAxisLayerCount : 1;
  return axisLayerCount * maxLabelBoxGirth;
};

/** @internal */
export function getPosition(
  parentDimension: Dimensions,
  { chartDimensions }: { chartDimensions: Dimensions },
  chartMargins: PerSideDistance,
  { axisTitle, axisPanelTitle, tickLine, tickLabel }: AxisStyle,
  { title, position, hide, timeAxisLayerCount }: AxisSpec,
  { maxLabelBboxHeight, maxLabelBboxWidth }: TickLabelBounds,
  smScales: SmallMultipleScales,
  { top: cumTopSum, bottom: cumBottomSum, left: cumLeftSum, right: cumRightSum }: PerSideDistance,
  verticalAxisCount: number,
) {
  const tickDimension = shouldShowTicks(tickLine, hide) ? tickLine.size + tickLine.padding : 0;
  const labelPaddingSum = tickLabel.visible ? innerPad(tickLabel.padding) + outerPad(tickLabel.padding) : 0;
  const titleDimension = title ? getTitleDimension(axisTitle) : 0;
  const vertical = isVerticalAxis(position);
  const scaleBand = vertical ? smScales.vertical : smScales.horizontal;
  const panelTitleDimension = hasSMDomain(scaleBand) ? getTitleDimension(axisPanelTitle) : 0;
  const maxLabelBboxGirth = tickLabel.visible ? (vertical ? maxLabelBboxWidth : maxLabelBboxHeight) : 0;
  const layerGrith = getAllAxisLayersGirth(timeAxisLayerCount, maxLabelBboxGirth, !vertical);
  const shownLabelSize = vertical ? Math.min(parentDimension.width / (verticalAxisCount * 2), layerGrith) : layerGrith;
  const parallelSize = labelPaddingSum + shownLabelSize + tickDimension + titleDimension + panelTitleDimension;
  return {
    leftIncrement: position === Position.Left ? parallelSize + chartMargins.left : 0,
    rightIncrement: position === Position.Right ? parallelSize + chartMargins.right : 0,
    topIncrement: position === Position.Top ? parallelSize + chartMargins.top : 0,
    bottomIncrement: position === Position.Bottom ? parallelSize + chartMargins.bottom : 0,
    dimensions: {
      left:
        position === Position.Left
          ? chartMargins.left + cumLeftSum
          : chartDimensions.left + (position === Position.Right ? chartDimensions.width + cumRightSum : 0),
      top:
        position === Position.Top
          ? chartMargins.top + cumTopSum
          : chartDimensions.top + (position === Position.Bottom ? chartDimensions.height + cumBottomSum : 0),
      width: vertical ? parallelSize : chartDimensions.width,
      height: vertical ? chartDimensions.height : parallelSize,
    },
  };
}

/** @internal */
export function shouldShowTicks({ visible }: AxisStyle['tickLine'], axisHidden: boolean): boolean {
  return !axisHidden && visible;
}

/** @internal */
export interface AxisGeometry {
  anchorPoint: Point;
  size: Size;
  parentSize: Size;
  axis: {
    id: AxisId;
    position: Position;
    panelTitle?: string; // defined later per panel
    secondary?: boolean; // defined later per panel
  };
  dimension: TickLabelBounds;
  visibleTicks: AxisTick[];
  maxLabelSize: number;
}

/** @internal */
export function getAxesGeometries(
  parentDimension: Dimensions,
  chartDims: { chartDimensions: Dimensions; leftMargin: number },
  { chartPaddings, chartMargins, axes: sharedAxesStyle }: Theme,
  axisSpecs: Map<AxisId, AxisSpec>,
  axesStyles: Map<AxisId, AxisStyle | null>,
  smScales: SmallMultipleScales,
  visibleTicksSet: Map<AxisId, Projection>,
): AxisGeometry[] {
  const verticalAxesCount = [...axisSpecs.values()].reduce((count, spec) => {
    return count + (isVerticalAxis(spec.position) ? 1 : 0);
  }, 0);
  const panel = getPanelSize(smScales);
  return [...visibleTicksSet].reduce(
    (acc: PerSideDistance & { geoms: AxisGeometry[] }, [axisId, { ticks, labelBox, scale }]: [AxisId, Projection]) => {
      const axisSpec = axisSpecs.get(axisId);
      if (axisSpec) {
        const vertical = isVerticalAxis(axisSpec.position);
        const axisStyle = axesStyles.get(axisId) ?? sharedAxesStyle;
        const { dimensions, topIncrement, bottomIncrement, leftIncrement, rightIncrement } = getPosition(
          parentDimension,
          chartDims,
          chartMargins,
          axisStyle,
          axisSpec,
          labelBox,
          smScales,
          acc,
          verticalAxesCount,
        );
        acc.top += topIncrement;
        acc.bottom += bottomIncrement;
        acc.left += leftIncrement;
        acc.right += rightIncrement;
        acc.geoms.push({
          axis: { id: axisSpec.id, position: axisSpec.position },
          anchorPoint: { x: dimensions.left, y: dimensions.top },
          dimension: labelBox,
          visibleTicks: ticks,
          parentSize: { height: dimensions.height, width: dimensions.width },
          size: {
            width: labelBox.isHidden ? 0 : vertical ? dimensions.width : panel.width,
            height: labelBox.isHidden ? 0 : vertical ? panel.height : dimensions.height,
          },
          maxLabelSize:
            vertical && !isContinuousScale(scale)
              ? parentDimension.width / (verticalAxesCount * 2)
              : isContinuousScale(scale)
                ? Infinity
                : scale.step,
        });
      } else {
        throw new Error(`Cannot compute scale for axis spec ${axisId}`); // todo move this feedback as upstream as possible
      }
      return acc;
    },
    { geoms: [], top: 0, bottom: chartPaddings.bottom, left: chartDims.leftMargin, right: chartPaddings.right },
  ).geoms;
}
