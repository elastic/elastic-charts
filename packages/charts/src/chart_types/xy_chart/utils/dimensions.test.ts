/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { TickLabelBounds } from './axis_utils';
import { computeChartDimensions } from './dimensions';
import { AxisSpec } from './specs';
import { ChartType } from '../..';
import { SpecType } from '../../../specs/constants';
import { Position } from '../../../utils/common';
import { Margins } from '../../../utils/dimensions';
import { AxisId } from '../../../utils/ids';
import { LIGHT_THEME } from '../../../utils/themes/light_theme';
import { LegendStyle } from '../../../utils/themes/theme';
import { AxesTicksDimensions } from '../state/selectors/compute_axis_ticks_dimensions';

describe('Computed chart dimensions', () => {
  const parentDim = {
    width: 100,
    height: 100,
    top: 0,
    left: 0,
  };
  const chartMargins: Margins = {
    left: 10,
    right: 10,
    top: 10,
    bottom: 10,
  };
  const chartPaddings: Margins = {
    left: 10,
    right: 10,
    top: 10,
    bottom: 10,
  };

  const axis1Dims: TickLabelBounds = {
    maxLabelBboxWidth: 10,
    maxLabelBboxHeight: 10,
    maxLabelTextWidth: 10,
    maxLabelTextHeight: 10,
    isHidden: false,
  };
  const axisLeftSpec: AxisSpec = {
    chartType: ChartType.XYAxis,
    specType: SpecType.Axis,
    id: 'axis_1',
    groupId: 'group_1',
    hide: false,
    showOverlappingTicks: false,
    showOverlappingLabels: false,
    position: Position.Left,
    tickFormat: (value: any) => `${value}`,
    timeAxisLayerCount: 0,
  };
  const legend: LegendStyle = {
    verticalWidth: 10,
    horizontalHeight: 10,
    spacingBuffer: 10,
    margin: 0,
    labelOptions: {
      maxLines: 1,
    },
  };
  const defaultTheme = LIGHT_THEME;
  const chartTheme = {
    ...defaultTheme,
    chartMargins,
    chartPaddings,
    axes: {
      ...defaultTheme.axes,
    },
    ...legend,
  };
  chartTheme.axes.axisTitle.fontSize = 10;
  chartTheme.axes.axisTitle.padding = 10;
  test('should be equal to parent dimension with no axis minus margins', () => {
    const axisDims: AxesTicksDimensions = new Map();
    const axisStyles = new Map();
    const axisSpecs: AxisSpec[] = [];
    const { chartDimensions } = computeChartDimensions(parentDim, chartTheme, axisDims, axisStyles, axisSpecs);
    expect(chartDimensions.left + chartDimensions.width).toBeLessThanOrEqual(parentDim.width);
    expect(chartDimensions.top + chartDimensions.height).toBeLessThanOrEqual(parentDim.height);
    expect(chartDimensions).toMatchSnapshot();
  });
  test('should be padded by a left axis', () => {
    // |margin|titleFontSize|titlePadding|maxLabelBboxWidth|tickPadding|tickSize|padding|
    // \10|10|10|10|10|10|10| = 70px from left
    const axisDims: AxesTicksDimensions = new Map();
    const axisStyles = new Map();
    const axisSpecs = [axisLeftSpec];
    axisDims.set('axis_1', axis1Dims);
    const { chartDimensions } = computeChartDimensions(parentDim, chartTheme, axisDims, axisStyles, axisSpecs);
    expect(chartDimensions.left + chartDimensions.width).toBeLessThanOrEqual(parentDim.width);
    expect(chartDimensions.top + chartDimensions.height).toBeLessThanOrEqual(parentDim.height);
    expect(chartDimensions).toMatchSnapshot();
  });
  test('should be padded by a right axis', () => {
    // |padding|tickSize|tickPadding|maxLabelBBoxWidth|titlePadding|titleFontSize\margin|
    // \10|10|10|10|10|10|10| = 70px from right
    const axisDims: AxesTicksDimensions = new Map();
    const axisStyles = new Map();
    const axisSpecs = [{ ...axisLeftSpec, position: Position.Right }];
    axisDims.set('axis_1', axis1Dims);
    const { chartDimensions } = computeChartDimensions(parentDim, chartTheme, axisDims, axisStyles, axisSpecs);
    expect(chartDimensions.left + chartDimensions.width).toBeLessThanOrEqual(parentDim.width);
    expect(chartDimensions.top + chartDimensions.height).toBeLessThanOrEqual(parentDim.height);
    expect(chartDimensions).toMatchSnapshot();
  });
  test('should be padded by a top axis', () => {
    // |margin|titleFontSize|titlePadding|maxLabelBboxHeight|tickPadding|tickSize|padding|
    // \10|10|10|10|10|10|10| = 70px from top
    const axisDims: AxesTicksDimensions = new Map();
    const axisStyles = new Map();
    const axisSpecs = [
      {
        ...axisLeftSpec,
        position: Position.Top,
      },
    ];
    axisDims.set('axis_1', axis1Dims);
    const { chartDimensions } = computeChartDimensions(parentDim, chartTheme, axisDims, axisStyles, axisSpecs);
    expect(chartDimensions.left + chartDimensions.width).toBeLessThanOrEqual(parentDim.width);
    expect(chartDimensions.top + chartDimensions.height).toBeLessThanOrEqual(parentDim.height);
    expect(chartDimensions).toMatchSnapshot();
  });
  test('should be padded by a bottom axis', () => {
    // |margin|titleFontSize|titlePadding|maxLabelBboxHeight|tickPadding|tickSize|padding|
    // \10|10|10|10|10|10|10| = 70px from bottom
    const axisDims: AxesTicksDimensions = new Map();
    const axisStyles = new Map();
    const axisSpecs = [
      {
        ...axisLeftSpec,
        position: Position.Bottom,
      },
    ];
    axisDims.set('axis_1', axis1Dims);
    const { chartDimensions } = computeChartDimensions(parentDim, chartTheme, axisDims, axisStyles, axisSpecs);
    expect(chartDimensions.left + chartDimensions.width).toBeLessThanOrEqual(parentDim.width);
    expect(chartDimensions.top + chartDimensions.height).toBeLessThanOrEqual(parentDim.height);
    expect(chartDimensions).toMatchSnapshot();
  });
  test('should not add space for axis when no spec for axis dimensions or axis is hidden', () => {
    const axisDims: AxesTicksDimensions = new Map();
    const axisStyles = new Map();
    const axisSpecs = [
      {
        ...axisLeftSpec,
        position: Position.Bottom,
      },
    ];
    axisDims.set('foo', axis1Dims);
    const chartDimensions = computeChartDimensions(parentDim, chartTheme, axisDims, axisStyles, axisSpecs);

    const expectedDims = {
      chartDimensions: {
        height: 60,
        width: 60,
        left: 20,
        top: 20,
      },
      leftMargin: 10,
    };

    expect(chartDimensions).toEqual(expectedDims);

    const hiddenAxisDims: AxesTicksDimensions = new Map();
    const hiddenAxisSpecs = new Map<AxisId, AxisSpec>();
    hiddenAxisDims.set('axis_1', axis1Dims);
    hiddenAxisSpecs.set('axis_1', {
      ...axisLeftSpec,
      hide: true,
      position: Position.Bottom,
    });
    const hiddenAxisChartDimensions = computeChartDimensions(parentDim, chartTheme, axisDims, axisStyles, axisSpecs);

    expect(hiddenAxisChartDimensions).toEqual(expectedDims);
  });
});
