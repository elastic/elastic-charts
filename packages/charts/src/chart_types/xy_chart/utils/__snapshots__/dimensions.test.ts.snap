// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Computed chart dimensions should be equal to parent dimension with no axis minus margins 1`] = `
{
  "height": 60,
  "left": 20,
  "top": 20,
  "width": 60,
}
`;

exports[`Computed chart dimensions should be padded by a bottom axis 1`] = `
{
  "height": 32,
  "left": 25,
  "top": 20,
  "width": 50,
}
`;

exports[`Computed chart dimensions should be padded by a left axis 1`] = `
{
  "height": 50,
  "left": 48,
  "top": 25,
  "width": 32,
}
`;

exports[`Computed chart dimensions should be padded by a right axis 1`] = `
{
  "height": 50,
  "left": 20,
  "top": 25,
  "width": 32,
}
`;

exports[`Computed chart dimensions should be padded by a top axis 1`] = `
{
  "height": 32,
  "left": 25,
  "top": 48,
  "width": 50,
}
`;
